net/tstun/wrap.go

```diff
@@ -160,14 +173,24 @@ func (t *Wrapper) SetDestIPActivityFuncs(m map[netaddr.IP]func()) {
 func (t *Wrapper) Close() error {
 	var err error
 	t.closeOnce.Do(func() {
-		// Other channels need not be closed: poll will exit gracefully after this.
 		close(t.closed)
-
+		close(t.bufferConsumed)
+		close(t.outbound)
 		err = t.tdev.Close()
 	})
 	return err
 }
 
+// isClosed reports whether t is closed.
+func (t *Wrapper) isClosed() bool {
+	select {
+	case <-t.closed:
```

Comment:
Is this as efficient as reading/writing an atomic? Maybe the loop that uses this is rare enough it doesn’t matter?

Reply:
It's very efficient. The compiler converts it to a call to runtime.selectnbrecv. That calls runtime.chanrecv, which does (for this input):

* two comparisons
* three atomic loads
* a call to runtime.typedmemclr, which clears exactly 0 bytes

So it's not zero cost (a couple of function calls, a few atomic loads), but it's cheap enough to be effectively free compared to everything else around it. Most relevant, there is no path in which the scheduler is involved; it is limited to only memory reads and comparisons.

Reply:
It's also not as good as it could be: https://github.com/golang/go/issues/47023